work17/index.js

```diff
@@ -0,0 +1,191 @@
+"use strict";
+const REQUEST_URL = "https://mocki.io/v1/d4a57e5a-8c84-4fee-aa05-70a1791c0d86";
+
+// 間違っているURLの場合↓
+// const REQUEST_URL = "https://myjson.dit.upm.es/api/bins/ほげほげajy3";
+// 空の配列の場合↓
+// const REQUEST_URL = "https://mocki.io/v1/242b685f-a3d7-45a8-aeca-0376bd495b89";
+// 503エラーの場合↓
+// const REQUEST_URL = "https://httpstat.us/503";
+
+const wrapDiv = document.querySelector(".slider_wrap");
+
+const renderLoading = () => {
+    const div = document.createElement("div");
+    const img = document.createElement("img");
+
+    div.classList = "loading-wrap";
+    img.src = "./img/loading-circle.gif";
+
+    wrapDiv.appendChild(div).appendChild(img);
+    document.body.appendChild(div);
+}
+
+const hideLoadingImg = () => {
+    document.querySelector(".loading-wrap").remove();
+}
+
+const init = async () => {
+    renderLoading();
+    let imgJsonData;
+    try {
+        imgJsonData = await getRequestData(REQUEST_URL);
+    } catch (error) {
+        renderErrorMessage(`${error.message}`)
+    } finally {
+        hideLoadingImg();
+    }
+
+    if (!imgJsonData.data.length) {
+        renderErrorMessage("JSONデータが空です。");
+        return;
+    }
+    renderSlideImg(imgJsonData.data);
+}
+
+const getRequestData = async REQUEST_URL => {
+    return new Promise(resolve => {
+        setTimeout(() => {
+            resolve(fetchSlideImgData(REQUEST_URL));
+        }, 3000);
+    })
+}
+
+const fetchSlideImgData = async () => {
+    const parameter = {
+        headers: {
+            'Content-Type': 'application/json'
+        }
+    }
+    const response = await fetch(REQUEST_URL, parameter);
+    if (!response.ok) {
+        renderErrorMessage(`${response.status}:${response.statusText}`);
+    } else {
+        const jsonData = response.json();
+        return jsonData;
+    }
+}
+
+const renderErrorMessage = errorMessage => {
+    const p = document.createElement("p");
+
+    p.textContent = errorMessage;
+    document.body.prepend(p);
+}
+
+const renderSlideImg = imagesDataArray => {
+    const slideDiv = document.createElement("div");
+    const ul = document.createElement("ul");
+    const fragment = document.createDocumentFragment();
+
+    ul.classList = "slider_item"
+    slideDiv.classList = "slider_slide";
+
+    for (let i = 0; i < imagesDataArray.length; i++) {
+        const li = document.createElement("li");
+        const img = document.createElement("img");
+
+        img.src = imagesDataArray[i].src;
+        li.classList = "slider_items";
+        if (!i) {
+            li.classList.add("is-display");
+        }
+        li.id = `js-sliderItems${i + 1}`;
+
+        li.appendChild(img);
+        fragment.appendChild(li);
+    }
+
+    document.body.appendChild(wrapDiv)
+        .appendChild(slideDiv).appendChild(ul).appendChild(fragment);
+    slideDiv.appendChild(renderPagingNumber());
+    renderSlideArrow();
+    clickedSliderEvent();
```

Comment:
IMO
I feel that it is too much of a burden to include the generation of the counter in `renderSlideImg `and the render and event definition for the arrow button.
Why not separate it into a separate function or execute it in `init()`?

Reply:
I tried both separating the functions into separate functions and putting them into the init function...

When separating the functions, the "clickedSliderEvent" function could be placed outside the "renderSlideImg" function, but since asynchronous processing was running, it was difficult to separate the "renderSlideArrow" function, so it was placed in the init function.

commit: [41aaf4e](https://github.com/yasu-pro/js-handson/pull/22/commits/41aaf4e052c9657826b5592fceadb9895b05f6ab)

Reply:
`createPagingNumber()` also should be executed outside of` renderSlideImg()`.
This is to make the function name match the actual job.
```js
slideDiv.appendChild(createPagingNumber());
```